hal/mbed_sleep_manager.c

```diff
@@ -19,13 +19,116 @@
 #include "mbed_error.h"
 #include <limits.h>
+#include <stdio.h>
 
 #if DEVICE_SLEEP
 
 // deep sleep locking counter. A target is allowed to deep sleep if counter == 0
 static uint16_t deep_sleep_lock = 0U;
 
-void sleep_manager_lock_deep_sleep(void)
+#ifdef MBED_SLEEP_TRACING_ENABLED
+
+// Length of the identifier extracted from the driver name to store for logging.
+#define IDENTIFIER_WIDTH 7
+// Number of drivers that can be stored in the structure
+#define STATISTIC_COUNT  10
+
+typedef struct sleep_statistic {
+    char identifier[IDENTIFIER_WIDTH];
+    uint8_t count;
+} sleep_statistic_t;
+
+static sleep_statistic_t sleep_stats[STATISTIC_COUNT];
+
+static const char* strip_path(const char* const filename)
+{
+    char *output = strrchr(filename, '/');
+
+    if (output != NULL) {
+        return output + 1;
+    }
+
+    output = strrchr(filename, '\\');
+
+    if (output != NULL) {
+        return output + 1;
+    }
+
+    return filename;
+}
+
+static size_t sleep_tracker_find_index(const char *const filename)
+{
+    char temp[IDENTIFIER_WIDTH];
+    strncpy(temp, filename, IDENTIFIER_WIDTH);
+    temp[IDENTIFIER_WIDTH - 1] = '\0';
+
+    // Search for the a driver matching the current name and return it's index
+    for (int i = 0; i < STATISTIC_COUNT; ++i) {
+        if (strcmp(sleep_stats[i].identifier, temp) == 0) {
+            return i;
+        }
+    }
+
+    // If no driver was found currently in the structure, find the first array
+    // index that hasn't been used by a driver and fill it, then return the
+    // index.
+    for (int i = 0; i < STATISTIC_COUNT; ++i) {
+        if (sleep_stats[i].identifier[0] == '\0') {
+            core_util_critical_section_enter();
+            strncpy(sleep_stats[i].identifier, temp, sizeof(temp));
+            core_util_critical_section_exit();
+
+            return i;
+        }
+    }
+
+    // Panic if there are no free indexes left to track with
+    error("No free indexes left to use in mbed stats tracker");
+
+    return -1;
+}
+
+static void sleep_tracker_print_stats(void)
+{
+    for (int i = 0; i < STATISTIC_COUNT; ++i) {
+        if (sleep_stats[i].count == 0) {
+            continue;
+        }
+
+        if (sleep_stats[i].identifier[0] == '\0') {
+            return;
+        }
+
+        printf("[id: %s, count: %u]\r\n", sleep_stats[i].identifier,
+                                          sleep_stats[i].count);
+    }
+}
+
+void sleep_tracker_lock(const char* const filename, int line)
+{
+    const char* const stripped_path = strip_path(filename);
+    size_t index = sleep_tracker_find_index(stripped_path);
+
+    core_util_atomic_incr_u8(&sleep_stats[index].count, 1);
+
+    printf("LOCK: %s, ln: %i, lock count: %u\r\n", stripped_path, line, deep_sleep_lock);
+}
+
+void sleep_tracker_unlock(const char* const filename, int line)
+{
+    const char* const stripped_path = strip_path(filename);
+    size_t index = sleep_tracker_find_index(stripped_path);
```

Comment:
It can be changed, though it's unlikely to cause problems at the moment. If a driver is unlocking sleep manager without first locking it 's going to cause an error() eventually via the unlock code which checks for underflows.

Reply:
It can be changed, though it's unlikely to cause problems at the moment. If a driver is unlocking sleep manager without first locking it 's going to cause an error() eventually via the unlock code which checks for underflows.

Reply:
Please make that change, if somebody tried to call unlock without locking it should be treated as an error. We should not add a new entry and go with it which might cause inadvertent behaviors.